From c49117ab45feb5db9cd6dd78e703797389220e47 Mon Sep 17 00:00:00 2001
From: Dave Airlie <airlied@redhat.com>
Date: Wed, 15 Jun 2016 10:17:44 +1000
Subject: [PATCH 3/6] blit: handle emulated alpha sources as well just in case.

We haven't seen this yet, but it is a potential problem we should fix.
---
 src/vrend_blitter.c  | 3 +++
 src/vrend_renderer.c | 3 ++-
 2 files changed, 5 insertions(+), 1 deletion(-)

diff --git a/src/vrend_blitter.c b/src/vrend_blitter.c
index 8bfc791..86c2061 100644
--- a/src/vrend_blitter.c
+++ b/src/vrend_blitter.c
@@ -600,6 +600,9 @@ void vrend_renderer_blit_gl(struct vrend_context *ctx,
 
    glBindTexture(src_res->target, src_res->id);
 
+   if (vrend_format_is_emulated_alpha(info->src.format))
+      glTexParameteri(src_res->target, GL_TEXTURE_SWIZZLE_R, GL_ALPHA);
+
    glTexParameteri(src_res->target, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
    glTexParameteri(src_res->target, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
    glTexParameteri(src_res->target, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
diff --git a/src/vrend_renderer.c b/src/vrend_renderer.c
index cc758ea..da7cbed 100644
--- a/src/vrend_renderer.c
+++ b/src/vrend_renderer.c
@@ -5498,7 +5498,8 @@ static void vrend_renderer_blit_int(struct vrend_context *ctx,
    if (info->src.box.depth != info->dst.box.depth)
       use_gl = true;
 
-   if (vrend_format_is_emulated_alpha(info->dst.format))
+   if (vrend_format_is_emulated_alpha(info->dst.format) ||
+       vrend_format_is_emulated_alpha(info->src.format))
       use_gl = true;
 
    if (use_gl) {
-- 
2.11.0

